src/main/java/fr/xephi/authme/listener/PlayerListener.java

```diff
@@ -190,46 +192,73 @@ public void onPlayerJoin(PlayerJoinEvent event) {
         management.performJoin(player);
     }
 
+    private void runOnJoinChecks(String name, String ip) throws FailedVerificationException {
+        // Fast stuff
+        onJoinVerifier.checkSingleSession(name);
+        onJoinVerifier.checkIsValidName(name);
+
+        // Get the auth later as this may cause the single session check to fail
+        // Slow stuff
+        final PlayerAuth auth = dataSource.getAuth(name);
+        final boolean isAuthAvailable = auth != null;
+        onJoinVerifier.checkKickNonRegistered(isAuthAvailable);
+        onJoinVerifier.checkAntibot(name, isAuthAvailable);
+        onJoinVerifier.checkNameCasing(name, auth);
+        onJoinVerifier.checkPlayerCountry(name, ip, isAuthAvailable);
+    }
+
     // Note #831: AsyncPlayerPreLoginEvent is not fired by all servers in offline mode
     // e.g. CraftBukkit does not fire it. So we need to run crucial things with PlayerLoginEvent.
     // Single session feature can be implemented for Spigot and CraftBukkit by canceling a kick
     // event caused by "logged in from another location". The nicer way, but only for Spigot, would be
     // to check in the AsyncPlayerPreLoginEvent. To support all servers, we use the less nice way.
 
+    private boolean isAsyncPlayerPreLoginEventCalled = false;
+
+    @EventHandler(priority = EventPriority.HIGHEST)
+    public void onAsyncPlayerPreLoginEvent(AsyncPlayerPreLoginEvent event) {
+        isAsyncPlayerPreLoginEventCalled = true;
+
+        final String name = event.getName();
+
+        if (validationService.isUnrestricted(name)) {
+            return;
+        }
```

Comment:
Here, like in the non-async method, would it be worth to check `event#result`? I don't think it's dramatic as it is right now, but we'd be more consistent with the same check.

Reply:
I think it would break some features like the single session kick message.